pkg/util/kubernetes/autoscalers/datadogexternal.go

```diff
@@ -84,75 +81,66 @@ func getMinRemainingRequestsTracker() *minTracker {
 	return minRemainingRequestsTracker
 }
 
-// IsRateLimitError is a helper function that checks if the received error is a rate limit error
-func IsRateLimitError(err error) bool {
-	if err == nil {
-		return false
-	}
-	return strings.Contains(err.Error(), "429 Too Many Requests")
-}
-
-// isUnprocessableEntityError is a helper function that checks if the received error is an unprocessable entity error
-func isUnprocessableEntityError(err error) bool {
-	if err == nil {
-		return false
-	}
-	return strings.Contains(err.Error(), "422 Unprocessable Entity")
-}
-
 // queryDatadogExternal converts the metric name and labels from the Ref format into a Datadog metric.
-// It returns the last value for a bucket of 5 minutes,
-func (p *Processor) queryDatadogExternal(ddQueries []string, timeWindow time.Duration) (map[string]Point, error) {
+// It should ALWAYS return either: (nil, err) or (map[string]Point, nil)
+// The former is used to signal a global error with the query, the latter is used to signal a successful query with potentially some errors per query.
+func (p *Processor) queryDatadogExternal(currentTime time.Time, ddQueries []string, timeWindow time.Duration) (map[string]Point, error) {
 	ddQueriesLen := len(ddQueries)
 	if ddQueriesLen == 0 {
 		log.Tracef("No query in input - nothing to do")
 		return nil, nil
 	}
 
-	query := strings.Join(ddQueries, ",")
-	currentTime := time.Now()
-	currentTimeUnix := time.Now().Unix()
-	seriesSlice, err := p.datadogClient.QueryMetrics(currentTime.Add(-timeWindow).Unix(), currentTimeUnix, query)
+	batchedQuery := strings.Join(ddQueries, ",")
+	currentTimeUnix := currentTime.Unix()
+	seriesSlice, err := p.datadogClient.QueryMetrics(currentTime.Add(-timeWindow).Unix(), currentTimeUnix, batchedQuery)
 	if err != nil {
-		if IsRateLimitError(err) {
+		apiErr := NewAPIError(err)
+		switch {
+		case apiErr.Code == RateLimitExceededAPIError:
 			ddRequests.Inc("rate_limit_error", le.JoinLeaderValue)
-		} else if isUnprocessableEntityError(err) {
+		case apiErr.Code == UnprocessableEntityAPIError:
```

Comment:
should we also log a debug line for this `case` ? Maybe a trace for the ratelimit error

Reply:
The two errors (429, 422) always have the same content and have dedicated metrics, the errors have debug logs because we don't actually what's the error or error content.